lib/rubygems/commands/rebuild_command.rb

```diff
@@ -0,0 +1,262 @@
+      options[:gemspec_file] = value
+    end
+
+    add_option "-C PATH", "Run as if gem build was started in <PATH> instead of the current working directory." do |value, options|
+      options[:build_path] = value
+    end
+  end
+
+  def arguments # :nodoc:
+    "GEM_NAME      gem name on gem server\n" \
+    "GEM_VERSION   gem version you are attempting to rebuild"
+  end
+
+  def description # :nodoc:
+    <<-EOF
+The rebuild command allows you to (attempt to) reproduce a build of a gem
+from a ruby gemspec.
+
+This command assumes the gemspec can be built with the `gem build` command.
+If you use either `gem build` or `rake build`/`rake release` to build/release
+a gem, it is a potential candidate.
+
+If the gem includes top-level files that change frequently (e.g. Gemfile.lock),
+it may require more effort to reproduce a build. For example, it might require
+more precisely matched versions of Ruby, RubyGems, or Bundler to be used.
+
+An example of reproducing a gem build:
+
+  $ pwd
+  /usr/home/puppy/test/rebuild
+  $ ls -a
+  ./  ../
+  $ git clone --branch v12.0.2 https://github.com/duckinator/okay.git
+  $ gem rebuild -C ./okay --gemspec okay.gemspec okay 12.0.2
+  Fetching okay-12.0.2.gem
+  Downloaded okay version 12.0.2 as /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem.
+    Successfully built RubyGem
+    Name: okay
+    Version: 12.0.2
+    File: okay-12.0.2.gem
+
+  Built at: 2023-08-31 21:39:02 EDT (1693532342)
+  Original build saved to:   /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
+  Reproduced build saved to: /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem
+  Working directory: ./okay
+
+  Hash comparison:
+    38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
+    38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem
+
+  SUCCESS - original and rebuild hashes matched
+    EOF
+  end
+
+  def usage # :nodoc:
+    "#{program_name} GEM_NAME GEM_VERSION"
+  end
+
+  def execute
+    gem_name, gem_version = get_gem_name_and_version
+
+    old_dir, new_dir = prep_dirs
+
+    gem_filename = "#{gem_name}-#{gem_version}.gem"
+    old_file = File.join(old_dir, gem_filename)
+    new_file = File.join(new_dir, gem_filename)
+
+    if options[:original_gem_file]
+      FileUtils.copy_file(options[:original_gem_file], old_file)
+    else
+      download_gem(gem_name, gem_version, old_file)
+    end
+
+    source_date_epoch = get_timestamp(old_file).to_s
+
+    if build_path = options[:build_path]
+      Dir.chdir(build_path) { build_gem(gem_name, source_date_epoch, new_file) }
+    else
+      build_gem(gem_name, source_date_epoch, new_file)
+    end
+
+    compare(source_date_epoch, old_file, new_file)
+  end
+
+  private
+
+  def sha256(file)
+    Digest::SHA256.hexdigest(File.read(file))
+  end
+
+  def get_timestamp(file)
```

Comment:
note to self: this is a sign we should embed the source date epoch used into the metadata of the built gemspec. additionally, ruby/rubygems versions would be helpful